# V3 empirical study — 5-day hourly calibration

**This notebook is the computational source of truth.** The PDF in this folder is written from the same `payload` dict.

| Item | Setting |
|------|---------|
| Models | GBM, GARCH, Heston, Merton, GARCH–Merton, Heston–Merton |
| Companies | SPY, AAPL, MSFT |
| Windows | 12 third-Friday monthly expiries, 21 Oct 2022 – 15 Sep 2023 |
| Evaluation | **five NYSE sessions ending the Friday before expiry (hourly 09:59–15:59)** |
| Calibration | **5-day** lookback, **hourly** rolling, 1-minute RTH data |
| LSM horizon | remaining time **to window Friday 15:59**, not listed expiry |
| LSM grid | Δt = **5 minute(s)**, n_steps = remaining RTH / Δt |
| Paths | `10000`, seed `42` |
| Sample | one nearest-ATM listed call per hourly timestamp, DTE 7–60 |

Green / Mark 1 = lowest RMSE%. Friday 15:59 is scheduled but LSM is skipped (`remaining_horizon = 0`).


## Methodology

- Shared sample: nearest-ATM, DTE 7–60, listed expiry, as-of each hourly *t*. All six models price the same contracts.
- No look-ahead: parameters at *t* use only equity returns and listed quotes with timestamp ≤ *t*.
- V3 Heston: option-implied NLS; Euler uses *v_t* then updates variance.
- LSM on the full path cloud. Start every path at the observed *S_t*.


In [1]:
import sys
from pathlib import Path
from IPython.display import display, HTML
ROOT = Path.cwd()
for cand in [ROOT, *ROOT.parents]:
    scripts = cand / 'V4-Models_result' / 'scripts'
    if scripts.exists():
        sys.path.insert(0, str(scripts))
        break
import run_v4_5d_hourly_return_based_study as study
RECOMPUTE = False
payload = study.run_or_load(recompute=RECOMPUTE)
print('cells', len(payload['cells']), 'failures', payload['meta']['failures'])

## Shared sample *n*


In [1]:
display(HTML('shared n'))

Window,Eval Friday,SPY n,AAPL n,MSFT n
Oct 2022 2022-10-21,2022-10-14,34,34,34
Nov 2022 2022-11-18,2022-11-11,34,34,34
Dec 2022 2022-12-16,2022-12-09,34,34,34
Jan 2023 2023-01-20,2023-01-13,34,34,34
Feb 2023 2023-02-17,2023-02-10,34,34,34
Mar 2023 2023-03-17,2023-03-10,34,34,34
Apr 2023 2023-04-21,2023-04-14,34,34,34
May 2023 2023-05-19,2023-05-12,34,34,34
Jun 2023 2023-06-16,2023-06-09,34,34,34
Jul 2023 2023-07-21,2023-07-14,34,34,34


## SPY


### Table SPY-1  ·  Oct 2022  2022-10-21  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['SPY|2022-10-21'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,74.58,70.42,-70.42,4
GBM,74.01,69.95,-69.95,3
GARCH,32.58,25.67,+2.76,1
Merton,262.87,235.20,+233.82,5
GARCH–Merton,72.90,68.67,-68.67,2


### Table SPY-2  ·  Nov 2022  2022-11-18  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['SPY|2022-11-18'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,99.87,99.87,-99.87,3
GBM,99.67,99.67,-99.67,2
GARCH,69.40,57.61,-57.31,1
Merton,234.03,196.04,+174.09,5
GARCH–Merton,99.93,99.93,-99.93,4


### Table SPY-3  ·  Dec 2022  2022-12-16  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['SPY|2022-12-16'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,3
GBM,100.00,100.00,-100.00,4
GARCH,88.40,84.60,-82.66,1
Merton,192.37,159.55,+147.44,5
GARCH–Merton,100.00,100.00,-100.00,2


### Table SPY-4  ·  Jan 2023  2023-01-20  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['SPY|2023-01-20'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,2
GBM,100.00,100.00,-100.00,3
GARCH,95.03,93.82,-93.82,1
Merton,163.83,134.29,+104.06,5
GARCH–Merton,100.00,100.00,-100.00,4


### Table SPY-5  ·  Feb 2023  2023-02-17  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['SPY|2023-02-17'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,3
GBM,100.00,100.00,-100.00,2
GARCH,83.79,78.09,-76.87,1
Merton,179.75,154.58,+135.69,5
GARCH–Merton,100.00,100.00,-100.00,4


### Table SPY-6  ·  Mar 2023  2023-03-17  ·  n = 34  ·  BEST = GARCH–Merton


In [1]:
display(HTML(study.emp._html_table(payload['tables']['SPY|2023-03-17'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,96.11,95.72,-95.72,3
GBM,95.77,95.37,-95.37,2
GARCH,118.20,96.44,+56.21,4
Merton,299.25,264.81,+262.24,5
GARCH–Merton,95.75,95.34,-95.34,1


### Table SPY-7  ·  Apr 2023  2023-04-21  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['SPY|2023-04-21'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,4
GBM,100.00,100.00,-100.00,3
GARCH,85.87,78.03,-5.16,1
Merton,209.44,178.03,+167.30,5
GARCH–Merton,99.93,99.93,-99.93,2


### Table SPY-8  ·  May 2023  2023-05-19  ·  n = 34  ·  BEST = GBM


In [1]:
display(HTML(study.emp._html_table(payload['tables']['SPY|2023-05-19'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,2
GBM,100.00,100.00,-100.00,1
GARCH,102.07,95.24,-2.07,4
Merton,197.37,163.95,+146.51,5
GARCH–Merton,100.00,100.00,-100.00,3


### Table SPY-9  ·  Jun 2023  2023-06-16  ·  n = 34  ·  BEST = MD-GBM


In [1]:
display(HTML(study.emp._html_table(payload['tables']['SPY|2023-06-16'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,1
GBM,100.00,100.00,-100.00,2
GARCH,101.38,90.80,-21.76,4
Merton,236.00,189.92,+164.48,5
GARCH–Merton,100.00,100.00,-100.00,3


### Table SPY-10  ·  Jul 2023  2023-07-21  ·  n = 34  ·  BEST = MD-GBM


In [1]:
display(HTML(study.emp._html_table(payload['tables']['SPY|2023-07-21'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,1
GBM,100.00,100.00,-100.00,2
GARCH,128.91,113.38,+24.53,4
Merton,241.60,197.57,+180.20,5
GARCH–Merton,100.00,100.00,-100.00,3


### Table SPY-11  ·  Aug 2023  2023-08-18  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['SPY|2023-08-18'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,2
GBM,100.00,100.00,-100.00,3
GARCH,78.24,70.81,-57.65,1
Merton,159.06,134.25,+105.80,5
GARCH–Merton,100.00,100.00,-100.00,4


### Table SPY-12  ·  Sep 2023  2023-09-15  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['SPY|2023-09-15'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,2
GBM,100.00,100.00,-100.00,3
GARCH,92.28,89.47,-86.79,1
Merton,142.96,113.83,+75.41,5
GARCH–Merton,100.00,100.00,-100.00,4


## AAPL


### Table AAPL-1  ·  Oct 2022  2022-10-21  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['AAPL|2022-10-21'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,61.31,58.72,-48.93,4
GBM,61.05,58.50,-48.72,3
GARCH,41.01,32.58,+5.91,1
Merton,254.24,236.24,+235.03,5
GARCH–Merton,60.52,58.06,-48.30,2


### Table AAPL-2  ·  Nov 2022  2022-11-18  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['AAPL|2022-11-18'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,96.80,96.46,-96.46,3
GBM,95.89,95.55,-95.55,2
GARCH,68.55,60.47,-60.47,1
Merton,232.52,208.27,+202.64,5
GARCH–Merton,97.12,96.67,-96.67,4


### Table AAPL-3  ·  Dec 2022  2022-12-16  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['AAPL|2022-12-16'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,92.26,91.75,-91.75,4
GBM,91.32,90.79,-90.79,3
GARCH,56.90,50.48,-23.51,1
Merton,254.50,231.07,+230.37,5
GARCH–Merton,88.90,88.30,-88.30,2


### Table AAPL-4  ·  Jan 2023  2023-01-20  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['AAPL|2023-01-20'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,99.72,99.72,-99.72,4
GBM,99.51,99.50,-99.50,3
GARCH,80.43,74.56,-72.30,1
Merton,195.11,162.93,+152.88,5
GARCH–Merton,99.42,99.42,-99.42,2


### Table AAPL-5  ·  Feb 2023  2023-02-17  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['AAPL|2023-02-17'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,97.63,97.49,-97.49,3
GBM,96.90,96.71,-96.71,2
GARCH,65.42,56.72,-37.23,1
Merton,278.14,242.18,+238.86,5
GARCH–Merton,98.46,98.32,-98.32,4


### Table AAPL-6  ·  Mar 2023  2023-03-17  ·  n = 34  ·  BEST = GARCH–Merton


In [1]:
display(HTML(study.emp._html_table(payload['tables']['AAPL|2023-03-17'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,77.99,75.74,-75.74,3
GBM,77.24,75.01,-75.01,2
GARCH,125.14,104.27,+69.59,4
Merton,362.53,319.29,+317.63,5
GARCH–Merton,77.02,74.55,-74.55,1


### Table AAPL-7  ·  Apr 2023  2023-04-21  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['AAPL|2023-04-21'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,84.91,82.77,-82.77,4
GBM,84.45,82.35,-82.35,3
GARCH,63.99,52.65,+21.45,1
Merton,204.41,182.41,+178.16,5
GARCH–Merton,83.84,81.82,-81.82,2


### Table AAPL-8  ·  May 2023  2023-05-19  ·  n = 34  ·  BEST = GBM


In [1]:
display(HTML(study.emp._html_table(payload['tables']['AAPL|2023-05-19'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,99.47,99.47,-99.47,2
GBM,99.19,99.17,-99.17,1
GARCH,101.12,96.96,-11.92,4
Merton,283.02,227.63,+211.22,5
GARCH–Merton,99.70,99.69,-99.69,3


### Table AAPL-9  ·  Jun 2023  2023-06-16  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['AAPL|2023-06-16'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,4
GBM,100.00,100.00,-100.00,3
GARCH,87.18,82.61,-76.82,1
Merton,174.05,153.86,+146.86,5
GARCH–Merton,99.97,99.97,-99.97,2


### Table AAPL-10  ·  Jul 2023  2023-07-21  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['AAPL|2023-07-21'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,3
GBM,100.00,100.00,-100.00,4
GARCH,82.42,75.42,-66.48,1
Merton,254.55,209.82,+190.89,5
GARCH–Merton,100.00,100.00,-100.00,2


### Table AAPL-11  ·  Aug 2023  2023-08-18  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['AAPL|2023-08-18'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,3
GBM,100.00,100.00,-100.00,4
GARCH,92.85,91.41,-91.41,1
Merton,228.80,193.95,+179.22,5
GARCH–Merton,100.00,100.00,-100.00,2


### Table AAPL-12  ·  Sep 2023  2023-09-15  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['AAPL|2023-09-15'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,3
GBM,100.00,100.00,-100.00,4
GARCH,99.76,99.75,-99.75,1
Merton,171.18,133.20,+126.40,5
GARCH–Merton,100.00,100.00,-100.00,2


## MSFT


### Table MSFT-1  ·  Oct 2022  2022-10-21  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['MSFT|2022-10-21'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,70.77,64.78,-37.92,3
GBM,70.40,64.50,-37.59,2
GARCH,52.40,36.58,+15.03,1
Merton,237.29,221.60,+221.60,5
GARCH–Merton,71.16,65.16,-38.26,4


### Table MSFT-2  ·  Nov 2022  2022-11-18  ·  n = 34  ·  BEST = GARCH–Merton


In [1]:
display(HTML(study.emp._html_table(payload['tables']['MSFT|2022-11-18'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,62.79,50.51,-15.38,3
GBM,62.41,50.04,-14.98,2
GARCH,174.12,136.24,+100.90,4
Merton,480.15,402.44,+401.10,5
GARCH–Merton,62.35,50.10,-16.77,1


### Table MSFT-3  ·  Dec 2022  2022-12-16  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['MSFT|2022-12-16'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,58.83,51.60,-51.60,3
GBM,57.59,50.63,-50.63,2
GARCH,56.21,44.10,+42.87,1
Merton,288.10,273.25,+273.25,5
GARCH–Merton,60.28,52.51,-52.51,4


### Table MSFT-4  ·  Jan 2023  2023-01-20  ·  n = 34  ·  BEST = GARCH–Merton


In [1]:
display(HTML(study.emp._html_table(payload['tables']['MSFT|2023-01-20'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,27.87,24.04,+3.55,2
GBM,27.91,24.09,+3.70,3
GARCH,131.19,104.69,+103.49,4
Merton,323.61,289.34,+289.34,5
GARCH–Merton,27.16,23.32,+2.60,1


### Table MSFT-5  ·  Feb 2023  2023-02-17  ·  n = 34  ·  BEST = GBM


In [1]:
display(HTML(study.emp._html_table(payload['tables']['MSFT|2023-02-17'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,96.79,89.73,-12.88,2
GBM,96.45,89.21,-12.30,1
GARCH,242.03,198.18,+133.45,4
Merton,530.35,461.60,+459.63,5
GARCH–Merton,97.16,90.30,-15.98,3


### Table MSFT-6  ·  Mar 2023  2023-03-17  ·  n = 34  ·  BEST = GBM


In [1]:
display(HTML(study.emp._html_table(payload['tables']['MSFT|2023-03-17'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,90.61,88.87,-88.87,2
GBM,89.94,88.14,-88.14,1
GARCH,144.78,112.36,+66.65,4
Merton,479.17,419.05,+418.94,5
GARCH–Merton,91.79,90.28,-90.28,3


### Table MSFT-7  ·  Apr 2023  2023-04-21  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['MSFT|2023-04-21'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,98.83,98.78,-98.78,3
GBM,98.57,98.51,-98.51,2
GARCH,51.49,40.24,-25.84,1
Merton,328.01,289.01,+284.97,5
GARCH–Merton,99.29,99.23,-99.23,4


### Table MSFT-8  ·  May 2023  2023-05-19  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['MSFT|2023-05-19'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,2
GBM,100.00,100.00,-100.00,3
GARCH,90.00,88.06,-88.06,1
Merton,181.23,139.98,+115.85,5
GARCH–Merton,100.00,100.00,-100.00,4


### Table MSFT-9  ·  Jun 2023  2023-06-16  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['MSFT|2023-06-16'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,3
GBM,100.00,100.00,-100.00,4
GARCH,97.45,97.30,-97.30,1
Merton,186.81,158.69,+143.94,5
GARCH–Merton,100.00,100.00,-100.00,2


### Table MSFT-10  ·  Jul 2023  2023-07-21  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['MSFT|2023-07-21'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,2
GBM,100.00,100.00,-100.00,3
GARCH,95.23,94.82,-94.82,1
Merton,225.67,194.21,+179.51,5
GARCH–Merton,100.00,100.00,-100.00,4


### Table MSFT-11  ·  Aug 2023  2023-08-18  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['MSFT|2023-08-18'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,2
GBM,100.00,100.00,-100.00,3
GARCH,99.99,99.99,-99.99,1
Merton,168.29,149.10,+129.04,5
GARCH–Merton,100.00,100.00,-100.00,4


### Table MSFT-12  ·  Sep 2023  2023-09-15  ·  n = 34  ·  BEST = GARCH


In [1]:
display(HTML(study.emp._html_table(payload['tables']['MSFT|2023-09-15'])))

Model,RMSE%,MAE%,Bias%,Ranking
MD-GBM,100.00,100.00,-100.00,2
GBM,100.00,100.00,-100.00,3
GARCH,100.00,100.00,-100.00,1
Merton,155.23,123.61,+84.65,5
GARCH–Merton,100.00,100.00,-100.00,4


## Summary S1


In [1]:
display(HTML('s1'))

Rank,Model,Mean RMSE%,Median RMSE%,# of 36 cells best
1,GBM,91.06,99.84,4
2,GARCH–Merton,91.18,99.93,4
3,MD-GBM,91.31,99.93,2
4,GARCH,93.77,91.14,26
5,Merton,249.87,233.27,0


## Findings

- On mean percentage RMSE across the 36 company×window cells, GBM is the best overall model (mean RMSE% = 91.06; 4 of 36 cells).

- Second overall is GARCH–Merton (mean RMSE% = 91.18; 4 cells).

- The ranking is not stable. Best-in-cell models across the 36 tables: GARCH, GARCH–Merton, GBM, MD-GBM.

- SPY: lowest mean RMSE% is GARCH (89.68); wins 8 of 12 windows.

- AAPL: lowest mean RMSE% is GARCH (80.40); wins 10 of 12 windows.

- MSFT: lowest mean RMSE% is GBM (83.61); wins 2 of 12 windows.

- No-jump models as a group have lower mean RMSE% (92.42) than the three jump models (170.53).

- These conclusions are conditional on the V3 filters, 5-day hourly calibration, nearest-ATM hourly sample, LSM with 10000 paths and 5-minute steps to Friday 15:59, and American-call quotes only. They are not a statement about European options or other tenors.
